### ISSUE TITLE
Add support for MSRPC based services

### ISSUE THREAD

**Issue by pepijnve** [hierynomus/smbj#82]
MSRPC services over SMB work by writing/reading data to/from an SMB named pipe. Adding support for this would enable things like remote share management, user management, etc.

I would propose adding some base classes to support this and then adapting MIDLC from the JCIFS project (http://jcifs.samba.org/src/midlc-0.6.1.tar.gz) to generate compatible classes. The adapted version of MIDLC would have to be GPLv2 licensed since the original code for IDLC falls under that license. The generated code is not GPLv2 licensed so I don't think this is an issue.

Is there any intrest in this feature?

**Comment by hierynomus** [comment]
Hi Pepijn,

I'm very allergic to GPLv2 licenses as we use SMBJ commercially. So my
preference is to keep them far away to prevent any confusion.

Adding base classes to support RPC over named pipes makes sense.

2017-06-15 10:07 GMT+02:00 Pepijn Van Eeckhoudt <notifications@github.com>:

**Comment by pepijnve** [comment]
Understood; same here.

I would suggest making modified MIDLC version a separate project; not part of SMBJ itself. SMBJ stays ASLv2. You would use MIDLC to generate request/response classes and the service client based on the MIDL files provided by Microsoft (see https://msdn.microsoft.com/en-us/library/cc247293.aspx for an example). The generated classes depend on runtime classes provided by SMBJ (e.g., a base class for an MSRPC service, encoding/decoding support classes, etc.).

The alternative is that you hand code all these classes, but that would just take a lot more time.

**Comment by hierynomus** [comment]
@pepijnve the guys from rapid7, especially @pmiseiko-r7 have started this feat in a separate repo https://github.com/rapid7/smbj-rpc.

**Comment by pepijnve** [comment]
Thanks for the pointer @hieronymous. That project Deens to be taking a slightly different approach than I had in mind where the messages are all being hand coded. Personally I think some kind of midl based code generator approach is going to be much more useful since that allows extra RPC services to be added in a much easier/quicker way once the base layer has been developed. But it's nice to see some activity in this area and might be a useful source of inspiration.

**Comment by pepijnve** [comment]
Just FYI, I started work on an IDL parser. First step in generation code based off of IDL files. As soon as I have something somewhat working I'll push this out to a feature branch so others can help if they want to.

**Comment by jbakermk** [comment]
@pepijnve Please can you tell me if you've done any work that would allow a [NetrLogonSamLogon](https://msdn.microsoft.com/en-us/library/cc237251.aspx) call through smbj? And if not, can you point me at the relevant code that would act as a starting point?

Thanks

**Comment by pepijnve** [comment]
@jbakermk no, what I'm working on is a code generator. Still in a very early state.
In the meantime the [smbj-rpc](https://github.com/rapid7/smbj-rpc) project is building RPC clients by hand. They seem to have support for SAM already. See [SecurityAccountManagerService.java](https://github.com/rapid7/smbj-rpc/blob/master/src/main/java/com/rapid7/client/dcerpc/mssamr/SecurityAccountManagerService.java). Not sure how complete this is and if it contains the request you're looking for.

**Comment by jbakermk** [comment]
@pepijnve Thanks!
